File: DemoBot/src/main/java/frc/robot/commands/ArcadeDrive.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import java.util.function.Supplier;

import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.robot.subsystems.Drivetrain;

public class ArcadeDrive extends CommandBase {
  Drivetrain drivetrain;
  Supplier<Double> xSupplier;
  Supplier<Double> zSupplier;
  /** Creates a new ArcadeDrive. */
  public ArcadeDrive(Drivetrain drive, Supplier<Double> xaxisSupplier, Supplier<Double> zaxisSupplier) {
    // Use addRequirements() here to declare subsystem dependencies.
    drivetrain = drive;
    xSupplier = xaxisSupplier;
    zSupplier = zaxisSupplier;
    addRequirements(drivetrain);
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {}

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    drivetrain.betterArcadeDrive(xSupplier.get(), zSupplier.get());
    
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {}

  // Returns true when the command should end.
  @Override
  public boolean isFinished() {
    return false;
  }
}

File: DemoBotCPP/src/main/cpp/commands/ArcadeDrive.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/ArcadeDrive.h"
#include "subsystems/Drivetrain.h"

ArcadeDrive::ArcadeDrive(Drivetrain* drivetrain, std::function<double(void)> forward, std::function<double(void)> rotation) {
  // Use addRequirements() here to declare subsystem dependencies.
  drive = drivetrain;
  x = forward;
  z = rotation;

  AddRequirements(drivetrain);
}

// Called when the command is initially scheduled.
void ArcadeDrive::Initialize() {}

// Called repeatedly when this Command is scheduled to run
void ArcadeDrive::Execute() {
  drive->SetSpeed(x() + z(), x() - z());
}

// Called once the command ends or is interrupted.
void ArcadeDrive::End(bool interrupted) {}

// Returns true when the command should end.
bool ArcadeDrive::IsFinished() {
  return false;
}
